### PR TITLE
fix: Do not panic for invalid lineage in transforms

### LINKED ISSUE #3266: Panic with DESC sort in aggregate without braces
### What happened?

When the following query is attempted to be compiled to sql, (first found via `prql_python`) the rust compiler panics with:
```
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', crates/prql-compiler/src/semantic/resolver/transforms.rs:528:54
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "xxx.py", line 51, in <module>
    sql = prql.compile(prql_query)
pyo3_runtime.PanicException: called `Option::unwrap()` on a `None` value
```

It does work without the `-`, and with `{-val}` instead.

### PRQL input

```elm
from tbl
group id (
    sort -val
    take 10
)
```

### SQL output

```SQL
-- Panic
```

### Expected SQL output

```SQL
WITH table_0 AS (
  SELECT
    *,
    ROW_NUMBER() OVER (
      PARTITION BY id
      ORDER BY
        val DESC
    ) AS _expr_0
  FROM
    tbl
)
SELECT
  *
FROM
  table_0
WHERE
  _expr_0 <= 10

-- Or an error about not wrapping the order in braces / invalid character `-` whichever is more correct for prql
```

### MVCE confirmation

- [X] Minimal example
- [X] New issue

### Anything else?

I tried looking at other issues but none of the panic ones mention this scenario.
#3111 might be related? But I have no idea as this is my first time using prql

### PR DESCRIPTION
**Pull request by davidot**
Makes the case in #3266 not panic.
But it gives a horrible error code dumping the ast? node or something like that.

Also this is essentially my first rust usage, so feel free to point out any rust things I'm doing wrong

Fixes #3266

The error with the input from the issue:

```
Error: "expected Expr { kind: Func(Func { name_hint: None, return_ty: None, body: Expr { kind: Func(Func { name_hint: None, return_ty: None, body: Expr { kind: Func(Func { name_hint: Some(Ident { path: [\"std\"], name: \"sub\" }), return_ty: Some(Ty(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") })]), name: None })), body: Expr { kind: Internal(\"std.sub\"), span: Some(1:1145-1161), alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }, params: [FuncParam { name: \"left\", ty: Some(Ty(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") })]), name: None })), default_value: None }, FuncParam { name: \"right\", ty: Some(Ty(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") })]), name: None })), default_value: None }], named_params: [], args: [Expr { kind: Func(Func { name_hint: Some(Ident { path: [\"std\"], name: \"sort\" }), return_ty: Some(Ty(Ty { kind: Array(Ty { kind: Tuple([Wildcard(Some(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"bool\"), Ty { kind: Primitive(Bool), name: Some(\"bool\") }), (Some(\"text\"), Ty { kind: Primitive(Text), name: Some(\"text\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") }), (Some(\"time\"), Ty { kind: Primitive(Time), name: Some(\"time\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (None, Ty { kind: Singleton(Null), name: None })]), name: Some(\"scalar\") }))]), name: Some(\"tuple_of_scalars\") }), name: Some(\"relation\") })), body: Expr { kind: Internal(\"sort\"), span: Some(1:2861-2874), alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }, params: [FuncParam { name: \"by\", ty: Some(Ty(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"bool\"), Ty { kind: Primitive(Bool), name: Some(\"bool\") }), (Some(\"text\"), Ty { kind: Primitive(Text), name: Some(\"text\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") }), (Some(\"time\"), Ty { kind: Primitive(Time), name: Some(\"time\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (None, Ty { kind: Singleton(Null), name: None }), (Some(\"tuple_of_scalars\"), Ty { kind: Tuple([Wildcard(Some(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"bool\"), Ty { kind: Primitive(Bool), name: Some(\"bool\") }), (Some(\"text\"), Ty { kind: Primitive(Text), name: Some(\"text\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") }), (Some(\"time\"), Ty { kind: Primitive(Time), name: Some(\"time\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (None, Ty { kind: Singleton(Null), name: None })]), name: Some(\"scalar\") }))]), name: Some(\"tuple_of_scalars\") })]), name: None })), default_value: None }, FuncParam { name: \"tbl\", ty: Some(Ty(Ty { kind: Array(Ty { kind: Tuple([Wildcard(Some(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"bool\"), Ty { kind: Primitive(Bool), name: Some(\"bool\") }), (Some(\"text\"), Ty { kind: Primitive(Text), name: Some(\"text\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") }), (Some(\"time\"), Ty { kind: Primitive(Time), name: Some(\"time\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (None, Ty { kind: Singleton(Null), name: None })]), name: Some(\"scalar\") }))]), name: Some(\"tuple_of_scalars\") }), name: Some(\"relation\") })), default_value: None }], named_params: [], args: [Expr { kind: Ident(Ident { path: [\"_param\"], name: \"_partial_311\" }), span: None, alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }, Expr { kind: Ident(Ident { path: [\"_param\"], name: \"_partial_317\" }), span: None, alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }], env: {} }), span: Some(0:24-28), alias: None, id: Some(317), target_id: None, target_ids: [], ty: Some(Ty { kind: Function(Some(TyFunc { args: [Some(Ty { kind: Array(Ty { kind: Tuple([Wildcard(Some(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"bool\"), Ty { kind: Primitive(Bool), name: Some(\"bool\") }), (Some(\"text\"), Ty { kind: Primitive(Text), name: Some(\"text\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") }), (Some(\"time\"), Ty { kind: Primitive(Time), name: Some(\"time\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (None, Ty { kind: Singleton(Null), name: None })]), name: Some(\"scalar\") }))]), name: Some(\"tuple_of_scalars\") }), name: Some(\"relation\") })], return_ty: Some(Ty { kind: Array(Ty { kind: Tuple([Wildcard(Some(Ty { kind: Union([(Some(\"int\"), Ty { kind: Primitive(Int), name: Some(\"int\") }), (Some(\"float\"), Ty { kind: Primitive(Float), name: Some(\"float\") }), (Some(\"bool\"), Ty { kind: Primitive(Bool), name: Some(\"bool\") }), (Some(\"text\"), Ty { kind: Primitive(Text), name: Some(\"text\") }), (Some(\"date\"), Ty { kind: Primitive(Date), name: Some(\"date\") }), (Some(\"time\"), Ty { kind: Primitive(Time), name: Some(\"time\") }), (Some(\"timestamp\"), Ty { kind: Primitive(Timestamp), name: Some(\"timestamp\") }), (None, Ty { kind: Singleton(Null), name: None })]), name: Some(\"scalar\") }))]), name: Some(\"tuple_of_scalars\") }), name: Some(\"relation\") }) })), name: None }), lineage: None, needs_window: false, flatten: false }, Expr { kind: Ident(Ident { path: [], name: \"val\" }), span: Some(0:30-33), alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }], env: {} }), span: None, alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }, params: [], named_params: [], args: [], env: {} }), span: None, alias: None, id: None, target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false }, params: [FuncParam { name: \"_partial_317\", ty: None, default_value: None }], named_params: [], args: [], env: {} }), span: None, alias: None, id: Some(313), target_id: None, target_ids: [], ty: None, lineage: None, needs_window: false, flatten: false } to have table type"
```

### CODE REVIEW

File: crates/prql-compiler/src/semantic/resolver/transforms.rs [hunk 1]
```diff
@@ -553,7 +552,7 @@ impl TransformCall {
                 // pipeline's body is resolved, just use its type
                 let Func { body, .. } = pipeline.kind.as_func().unwrap().as_ref();
 
-                body.lineage.clone().unwrap()
+                ty_frame_or_default(body)?
```
Comment (max-sixty) [inline crates/prql-compiler/src/semantic/resolver/transforms.rs]: Given how bad the error message is, let's just point to the issue? 

```suggestion
                ty_frame_or_default(body).map_err(anyhow!("Invalid lineage. See #3266 for more details."))?
```

File: crates/prql-compiler/src/semantic/resolver/transforms.rs [hunk 2]
```diff
@@ -1068,4 +1067,28 @@ mod tests {
...
+    fn test_invalid_lineage_in_transform() {
+        let result = parse_resolve_and_lower(
+            "
+        from tbl
+        group id (
+            sort -val
+        )
+        ",
+        );
+        assert!(result.is_err());
```
Comment (max-sixty) [inline crates/prql-compiler/src/semantic/resolver/transforms.rs]: Could we move this into `test_bad_error_messages.rs` and snapshot the error message? (If there's no difference between the examples, fine to keep only one)

**Review by max-sixty — COMMENTED** [review]
Nice work! Impressive for a first rust PR!

I asked for a couple of small changes. Let me know if you need any help with snapshots; some instructions here: https://prql-lang.org/book/project/contributing/development.html

Would be nice to add a Changelog but no obligation.

**Comment by max-sixty** [comment]
(the test failure is from needing to run `cargo fmt`)

**Comment by max-sixty** [comment]
@davidot no rush at all, but wanted to check you had the support you needed. 

If you're less comfortable about moving the tests and creating a snapshot, run `cargo fmt` on this, we can merge, and then I can do the final part in another PR.

**Comment by davidot** [comment]
Only got around to it now, let me know if the tests are good like this or should I add insta tests as well?

File: crates/prql-compiler/src/semantic/resolver/transforms.rs [hunk 3]
```diff
@@ -1068,4 +1071,27 @@ mod tests {
...
+    //     let result = parse_resolve_and_lower(
+    //         "
+    //     from tbl
+    //     window rows:-3..3 (
+    //         sort -val
+    //     )
+    //     ",
+    //     );
+    //     assert!(result.is_err());
+    // }
```
Comment (max-sixty) [inline crates/prql-compiler/src/semantic/resolver/transforms.rs]:
```suggestion
```

**Comment by max-sixty** [comment]
> Only got around to it now, let me know if the tests are good like this or should I add insta tests as well?

This is fine! Same purpose (insta is easier when doing lots of tests but I guess copy & paste maybe easier for a one-off...)

We can merge on green, thank you!

File: crates/prql-compiler/src/semantic/resolver/transforms.rs (hunk 3 above)
Comment (davidot) [reply crates/prql-compiler/src/semantic/resolver/transforms.rs]: Whoops, didn't check the diff thoroughly enough

**Pull request merged**